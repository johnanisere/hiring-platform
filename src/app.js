'use strict';
exports.__esModule = true;
var http_errors_1 = require('http-errors');
var express_1 = require('express');
var cors_1 = require('cors');
var compression_1 = require('compression');
var morgan_1 = require('morgan');
var express_graphql_1 = require('express-graphql');
var path_1 = require('path');
var mongoose_1 = require('mongoose');
var index_1 = require('./routes/index');
var schema_1 = require('./schema');
var app = express_1['default']();
// Setup Request logging
var logFormat = process.env.NODE_ENV === 'production' ? 'combined' : 'dev';
app.use(
  morgan_1['default'](logFormat, {
    skip: function(_req, res) {
      if (process.env.NODE_ENV === 'test') {
        return true;
      }
      return res.statusCode < 400;
    },
    stream: process.stderr,
  }),
);
app.use(
  morgan_1['default'](logFormat, {
    skip: function(_req, res) {
      if (process.env.NODE_ENV === 'test') {
        return true;
      }
      return res.statusCode >= 400;
    },
    stream: process.stdout,
  }),
);
app.disable('x-powered-by');
app.use(compression_1['default']());
app.use(cors_1['default']());
app.use(express_1['default'].json());
app.use(express_1['default'].urlencoded({ extended: false }));
app.use('/api', index_1['default']);
mongoose_1['default'].connect(
  'mongodb://localhost:27017/test',
  { useNewUrlParser: true },
  function() {
    console.log('Database connected succesfully');
  },
);
var connection = mongoose_1['default'].connection;
connection.once('open', function() {
  console.log('MongoDB database connection established successfully');
});
app.use(
  '/graphql',
  express_graphql_1['default']({
    schema: schema_1['default'],
    graphiql: true,
  }),
);
if (process.env.NODE_ENV === 'production') {
  app.use(
    express_1['default'].static(
      path_1['default'].join(__dirname, '../', 'client/build'),
    ),
  );
  app.get('/*', function(_req, res) {
    res.sendFile(
      path_1['default'].join(__dirname, '../', 'client/build/index.html'),
    );
  });
}
// catch 404 and forward to error handler
app.use(function(_req, _res, next) {
  next(http_errors_1['default'](404));
});
// error handler
app.use(function(err, req, res) {
  // set locals, only providing error in development
  res.locals.message = err.message;
  res.locals.error = req.app.get('env') === 'development' ? err : {};
  // render the error page
  res.status(err.status || 500);
  res.render('error');
});
exports['default'] = app;
