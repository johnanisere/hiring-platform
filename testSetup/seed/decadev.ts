import User from '../../src/models/User';

const users = [
  {
    email: 'doe@example.com',
    name: 'John Doe',
    phone: '08074583218',
    password: 'mysecret2',
    role: 'admin',
    profilePhoto:
      'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcSIcqjP8a8EfGrtkCX0CJXJ4HRltM5u8Jdhkz_TneWdb4rfPsMi',
  },

  {
    email: 'decadev@example.com',
    name: 'Jane Mary',
    phone: '08074583218',
    password: 'mysecret2',
    role: 'dev',
    profilePhoto:
      'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcSIcqjP8a8EfGrtkCX0CJXJ4HRltM5u8Jdhkz_TneWdb4rfPsMi',
    pod: 'android',
  },
];

export const seedUsers = async () => {
  try {
    const allUsers = users.map(async user => {
      const newUser = new User(user);
      return newUser.save();
    });

    const res = await Promise.all(allUsers);
    return res;
  } catch (err) {
    return err;
  }
};

const seed = async () => {
  try {
    const res = await seedUsers();
    console.log(`Database has been seeded with ${res.length} users`);
  } catch (err) {
    console.log({ err });
  }
};

export default seed;
