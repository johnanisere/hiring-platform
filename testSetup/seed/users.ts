import User from '../../src/models/User';
import HiringPartner from '../../src/models/HiringPartner';

const users = [
  {
    email: 'doe@example.com',
    name: 'John Doe',
    phone: '08074583218',
    password: 'mysecret2',
    role: 'admin',
    profilePhoto:
      'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcSIcqjP8a8EfGrtkCX0CJXJ4HRltM5u8Jdhkz_TneWdb4rfPsMi',
  },

  {
    email: 'decadev-user@example.com',
    name: ' Test Decadev',
    phone: '08074583218',
    password: 'mysecret',
    role: 'dev',
    profilePhoto:
      'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcSIcqjP8a8EfGrtkCX0CJXJ4HRltM5u8Jdhkz_TneWdb4rfPsMi',
    pod: 'java',
  },
  {
    email: 'decadev-use2r@example.com',
    name: ' Test Decadev',
    phone: '08074583208',
    password: 'mysecret',
    role: 'dev',
    profilePhoto:
      'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcSIcqjP8a8EfGrtkCX0CJXJ4HRltM5u8Jdhkz_TneWdb4rfPsMi',
    pod: 'nodejs',
  },
];

const hirers = [
  {
    name: 'Dola Akingbola',
    email: 'dola@example.com',
    nameOfOrg: 'Twitter',
    designation: 'CTO',
    website: 'https://www.google.com',
    industry: 'Technology',
    phone: '08066589871',
    numberOfTalentsRequired: '6-10',
    deadline: 'Within 1 Month',
    verified: true,
    active: true,
    password: 'mysecret',
    interestLanguage: [],
    interviews: [],
  },
  {
    name: 'Test Hirer',
    email: 'hirer2@example.com',
    nameOfOrg: 'Paystack',
    designation: 'CTO',
    website: 'https://www.google.com',
    industry: 'Technology',
    phone: '08066559871',
    numberOfTalentsRequired: '6-10',
    deadline: 'Within 1 Month',
    verified: false,
    active: false,
    password: 'mysecret',
    interestLanguage: [],
    interviews: [],
  },
];

export const seedUsers = async () => {
  try {
    const allUsers = users.map(async user => {
      const newUser = new User(user);
      return newUser.save();
    });

    const res = await Promise.all(allUsers);
    return res;
  } catch (err) {
    return err;
  }
};

export const seedPartners = async () => {
  try {
    const allPartners = hirers.map(async partner => {
      const newPartner = new HiringPartner(partner);
      return await newPartner.save();
    });
    const res = await Promise.all(allPartners);
    return res;
  } catch (err) {
    return err;
  }
};

const seed = async () => {
  try {
    const res = await seedUsers();
    console.log(`Database has been seeded with ${res.length} users`);
    const partners = await seedPartners();
    console.log(`Database has been seeded with ${partners.length} partner(s)`);
  } catch (err) {
    console.log({ err });
  }
};

export default seed;
