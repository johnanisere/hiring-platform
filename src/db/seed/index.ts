import users from './data/users';
import User from '../../models/User';
import Cycle from '../../models/Cycle';
import Employment from '../../models/Employment';
import employments from './data/employments';

const cleanDb = async () => {
  try {
    await Cycle.deleteMany({});
    await User.deleteMany({});
    return console.log('succesfully cleared db');
  } catch (err) {
    console.log('Error: occured', err);
    return err;
  }
};

export const seedUsers = async () => {
  try {
    const allUsers = users.map(async user => {
      const newUser = await new User(user);
      return newUser.save();
    });
    const cycle = await new Cycle();
    await cycle.save();
    const res = await Promise.all(allUsers);
    return res;
  } catch (err) {
    return err;
  }
};

export const seedEmployments = async () => {
  try {
    const allEmployments = employments.map(async employment => {
      const newEmployment = await new Employment(employment);
      return newEmployment.save();
    });
    const res: any = await Promise.all(allEmployments);
    const dev: any = await User.findOne({ email: 'sheyiogundijo@gmail.com' });
    res.forEach((employment: any) => dev.employments.push(employment._id));
    await dev.save();
    return res;
  } catch (err) {
    return err;
  }
};

const seed = () => {
  cleanDb()
    .then(() => {
      return seedUsers();
    })
    .then(() => {
      return seedEmployments();
    })
    .then(res => {
      console.log(`Database has been seeded with ${res.length} users`);
    })
    .catch(err => {
      console.log({ err });
    });
};

export default seed;
