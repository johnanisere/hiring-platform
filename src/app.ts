import dotenv from 'dotenv';
import createError from 'http-errors';
import express from 'express';
import cors from 'cors';
import compression from 'compression';
import morgan from 'morgan';
import graphQLHTTP from 'express-graphql';
import path from 'path';
import seed from './db/seed';
import dbConnection from './db';
import apiRouter from './routes/index';
import usersRouter from './routes/users';
import schema from './schema';

dotenv.config();
const app = express();

dbConnection.once('open', function() {
  seed();
  console.log('MongoDB database connection established successfully');
});

// Setup Request logging
const logFormat = process.env.NODE_ENV === 'production' ? 'combined' : 'dev';

app.use(
  morgan(logFormat, {
    skip: function(_req, res) {
      if (process.env.NODE_ENV === 'test') {
        return true;
      }

      return res.statusCode < 400;
    },
    stream: process.stderr,
  }),
);

app.use(
  morgan(logFormat, {
    skip: function(_req, res) {
      if (process.env.NODE_ENV === 'test') {
        return true;
      }

      return res.statusCode >= 400;
    },
    stream: process.stdout,
  }),
);

app.disable('x-powered-by');
app.use(compression());
app.use(cors());
app.use(express.json());
app.use(express.urlencoded({ extended: false }));

app.use('/api', apiRouter);
app.use('/users', usersRouter);

app.use(
  '/graphql',
  graphQLHTTP({
    schema,
    graphiql: true,
  }),
);

if (process.env.NODE_ENV === 'production') {
  app.use(express.static(path.join(__dirname, '../', 'client/build')));
  app.get('/*', (_req, res) => {
    res.sendFile(path.join(__dirname, '../', 'client/build/index.html'));
  });
}

// catch 404 and forward to error handler
app.use(function(
  _req: express.Request,
  _res: express.Response,
  next: express.NextFunction,
) {
  next(createError(404));
});

// error handler
app.use(function(err: any, req: express.Request, res: express.Response) {
  // set locals, only providing error in development
  res.locals.message = err.message;
  res.locals.error = req.app.get('env') === 'development' ? err : {};

  // render the error page
  res.status(err.status || 500);
  res.render('error');
});

export default app;
