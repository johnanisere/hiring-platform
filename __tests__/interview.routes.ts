import request from 'supertest';
import app from '../src/app';

import seed from '../testSetup/seed/interviews';

const { connectMongoDB, disconnectMongoDB } = require('../testSetup/mongodb');

const credentials = {
  email: 'dola@example.com',
  password: 'mysecret',
};
const adminCredentials = {
  email: 'doe@example.com',
  password: 'mysecret2',
};

let interviewId: string;
let hirer;
let token: string;
let admin: any;

beforeAll(async () => {
  await connectMongoDB('interview-tests');

  await seed();

  hirer = await request(app)
    .post('/api/v1/hirer/login')
    .send(credentials);
  token = hirer.body.token;

  admin = await request(app)
    .post('/api/v1/users/login')
    .send(adminCredentials);
});

afterAll(() => disconnectMongoDB());

describe('interview route', () => {
  test('schedule interview', async () => {
    expect.assertions(3);

    const interview = await request(app)
      .post('/api/v1/interview/invite')
      .set('Authorization', `Bearer ${token}`)
      .send({
        hiringPartner: 'dola@example.com',
        decaDev: 'interviewed@example.com',
        location: 'Victoria Island',
        startTime: '10am',
        endTime: '11am',
        startDate: 'wed, 1st oct 2020',
        endDate: 'wed, 1st oct 2020',
        description:
          'This is to inform you that you have been shortlisted for an interview',
        nameOfOrg: 'Paystack',
      })
      .expect(res => {
        const { id, ...rest } = res.body.interviewData;

        expect(res.status).toBe(200);
        expect(res.body).toEqual(
          expect.objectContaining({
            interviewData: expect.objectContaining({
              hiringPartner: expect.any(String),
              decaDev: expect.any(String),
              location: expect.any(String),
              startTime: expect.any(String),
              endTime: expect.any(String),
              description: expect.any(String),
              id: expect.any(String),
              startDate: expect.any(String),
              endDate: expect.any(String),
              scheduled: expect.any(String),
            }),
            message: "Interview has been sent to Decadev's email",
          }),
        );

        expect(rest).toMatchSnapshot();
      });
    interviewId = interview.body.interviewData.id;
  });
  test('Gets all Interviews', async () => {
    expect.assertions(2);
    return request(app)
      .get('/api/v1/interview/get-interviews')
      .set('Authorization', `Bearer ${admin.body.token}`)
      .expect(res => {
        expect(res.status).toBe(200);
        expect(Object.keys(res.body)).toContain('allInterviews');
      });
  });
  test('Adds Reason for interview decline', async () => {
    expect.assertions(2);
    return await request(app)
      .put('/api/v1/interview/why-decline/')
      .send({
        interviewId: interviewId,
        declineReason: 'Time Conflict',
      })
      .expect(res => {
        expect(res.status).toBe(200);
        expect(res.body).toEqual(
          expect.objectContaining({
            message: 'Interview Invitation has been declined',
          }),
        );
      });
  });
});
