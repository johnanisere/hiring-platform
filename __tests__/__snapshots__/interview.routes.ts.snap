// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`interview route schedule interview 1`] = `
Object {
  "decaDev": "interviewed@example.com",
  "description": "This is to inform you that you have been shortlisted for an interview",
  "endDate": "wed, 1st oct 2020",
  "endTime": "11am",
  "hiringPartner": "dola@example.com",
  "location": "Victoria Island",
  "scheduled": "wed, 1st oct 2020",
  "startDate": "wed, 1st oct 2020",
  "startTime": "10am",
}
`;
